KCON-106 document commons source (#393)
Aiven-Open/cloud-storage-connectors-for-apache-kafka: commons/src/main/java/io/aiven/kafka/connect/common/source/README.md

# Source design strategy

This document describes the design for source connectors and how to think about implementing one.

## Overview

The `AbstractSourceTask` handles most of the issues around synchronization and communicating with the upstream connector framework.  Implementors of concrete SourceTasks need only concern themselves with reading data from the data storage and an iterator of SourceRecord objects from that.  Concrete implementation need only implement three (3) methods:

 * `SourceCommonConfig configure(Map<String, String> props)`
 * `Iterator<SourceRecord> getIterator(BackoffConfig config)`
 * `void closeResources()`

> In addition, implementations will want to implement an `OffsetManager.OffsetManagerEntry`.  This class maps the information about the stored object to information describing how for the processing has progressed.  In general, the values that are needed to locate the original object in the object store are placed in the 'key' and data describing the processing is placed in the data.  There are no restricitons on either key or data elements except that they must be natively serializable by the Kafka process.

Reviewer: ```suggestion
In addition, implementations will want to implement an `OffsetManager.OffsetManagerEntry`.  This class maps the information about the stored object to information describing how far the processing has progressed.  In general, the values that are needed to locate the original object in the object store are placed in the 'key' and data describing the processing is placed in the data.  There are no restrictions on either key or data elements except that they must be natively serializable by the Kafka process.
```

> In addition, implementations will want to implement an `OffsetManager.OffsetManagerEntry`.  This class maps the information about the stored object to information describing how for the processing has progressed.  In general, the values that are needed to locate the original object in the object store are placed in the 'key' and data describing the processing is placed in the data.  There are no restricitons on either key or data elements except that they must be natively serializable by the Kafka process.

Reviewer: ```suggestion
In addition, implementations will want to implement an `OffsetManager.OffsetManagerEntry`.  This class maps the information about the stored object to information describing how for the processing has progressed.  In general, the values that are needed to locate the original object in the object store are placed in the 'key' and data describing the processing is placed in the data.  There are no restricitons on either key or data elements except that they must be natively serializable by the Kafka process (see https://github.com/a0x8o/kafka/blob/master/connect/runtime/src/main/java/org/apache/kafka/connect/storage/OffsetStorageReaderImpl.java for more info).
```
wdyt if we provide this reference as well ?

> In addition, implementations will want to implement an `OffsetManager.OffsetManagerEntry`.  This class maps the information about the stored object to information describing how for the processing has progressed.  In general, the values that are needed to locate the original object in the object store are placed in the 'key' and data describing the processing is placed in the data.  There are no restricitons on either key or data elements except that they must be natively serializable by the Kafka process.

Reviewer: How about providing an example of how key and value looks like here ?

### Helper classes

The `Transformer` classes found in the source/input directory describe how to convert input streams into objects that the Kafka framework can handle.  The object read from the storage is converted into a stream of bytes and passed to a transformer.  The transformer then creates a stream of SchemaAndValue objects that are the values for the object, these will be sent to Kafka as values in SourceRecords.  The transformer also creates the key for the source record.  

Each SchemaAndValue in the stream from the transformer is associated with an OffsetManagerEntry. The OffsetManagerEntry, the key, and the value are all that is needed to construct a SourceRecord to be sent to the SourceTask in the SourceRecord Iterator.

The `ObjectDistributionStrategy` provides convenient methods to assign objects to tasks and should be used early in the processing of objects from storage to determine if the current task should handle them.

## General process

> The `configure` method is called first.  The map provided is the data from the configuration.  The AbstractSourceTask implementation should create an instance of SourceCommonConfig and return it.  While it is doing this it should verify that all the data necessary for the SourceTask to operate has been defined.

Reviewer: ```suggestion
The `configure` method of class ? is called first.  The map provided is the data from the configuration.  The AbstractSourceTask implementation should create an instance of SourceCommonConfig and return it.  While it is doing this it should verify that all the data necessary for the SourceTask to operate has been defined.
```
I think from S3SourceTask ?

 * `Optional<E> getEntry(final OffsetManagerKey key, final Function<Map<String, Object>, E> creator)`  - This method retrieves any data Kafka has about the object identified by the key.  If the data are found then the `creator` function is called to create the OffsetManagerEntry implementation.  Otherwise, an empty Optional is returned.
 * `void remove(final OffsetManagerKey key)` - removes the data from the local OffsetManager cache.  This is called when data has been set to Kafka for the record identified by the key. Subsequent calls to `getEntry` will reload the data from Kafka.
 * `void remove(final SourceRecord sourceRecord)` - removes the data from the local OffsetManager cache.  This is called when the SourceRecord is sent to Kafka.  Subsequent calls to `getEntry` will reload the data from Kafka.

>  * `void load(Collection<OffsetManagerKey> keys)` - loads the data associated with a collection of keys into the OffsetManager cache so that they can be processed without further calls to retrieve the offset data from Kafka.

Reviewer 2: This has should be  `void populateOffsetManager`

### OffsetManagerEntry

The `OffsetManagerEntry` is the base class for source specific `OffsetManager` data.  It contains and manages the key and data for the OffsetManager.  Every data source has a way to identify the data source within the system.  This information should be mapped into the OffsetManagerEntry key.  Every processing stream that converts data from the source representation to the final SourceRecord(s) has state that it needs to track.  The OffsetManagerEntry handles that information and makes it easy to store and retrieve it.  The key and the data are key/value maps where the key is a string and the value is an object.  The OffsetManagerEntry provides access to the map data via getter and setter methods and provides conversions to common data types. Implementations of OffsetManagerEntry should call the getters/setters to get/store data that will be written to the kafka storage.  The class may track any other data this is desired by the developers but only the data in the key and data maps will be preserved.

> When thinking about the OffsetManagerEntry they key should be a simple map of key/value pairs that describe the location of the object in the storage system.  For something like S3 this would be the bucket and object key, other systems have other coordinate systems.  The `getManagerKey` method should return the map.  It is important to note that all integral numbers are stored in the Kafka maps as Longs.  So all integral key values should be stored a Longs in the map.  When integer data values are retrieved the `getInteger(String)` method may be used to ensure that the Long is converted into an Integer.

Reviewer 2: Maybe add a second example so as to make it obvious this can be re-used for a lot of different types of connectors?

"For something like S3 this would be the bucket and object key, or for a database it would be table and primaryKey, other systems have other coordinate systems"

> When thinking about the OffsetManagerEntry they key should be a simple map of key/value pairs that describe the location of the object in the storage system.  For something like S3 this would be the bucket and object key, for a database it would be table and primaryKey, other systems have other coordinate systems.  The `getManagerKey` method should return the map.  It is important to note that all integral numbers are stored in the Kafka maps as Longs.  So all integral key values should be stored a Longs in the map.  When integer data values are retrieved the `getInteger(String)` method may be used to ensure that the Long is converted into an Integer.

Reviewer: ```suggestion
When thinking about the OffsetManagerEntry the key should be a simple map of key/value pairs that describe the location of the object in the storage system.  For something like S3 this would be the bucket and object key, for a database it would be table and primaryKey, other systems have other coordinate systems.  The `getManagerKey` method should return the map.  It is important to note that all integral numbers are stored in the Kafka maps as Longs.  So all integral key values should be stored a Longs in the map.  When integer data values are retrieved the `getInteger(String)` method may be used to ensure that the Long is converted into an Integer.
```
